c10/cuda/CUDAFunctions.cpp

```diff
@@ -205,18 +208,28 @@ cudaError_t GetDeviceCount(int* dev_count) {
 // call y = torch.empty(1, device=“cuda”) # CUDA context is created on cuda:0
 // ```
 #if CUDA_VERSION >= 12000
-thread_local int targetDeviceIndex = -1;
+thread_local DeviceIndex targetDeviceIndex = -1;
 
-cudaError_t GetDevice(int* device) {
+cudaError_t GetDevice(DeviceIndex* device) {
   if (targetDeviceIndex >= 0) {
     *device = targetDeviceIndex;
     return cudaSuccess;
   }
-  return cudaGetDevice(device);
+  int tmp_device = -1;
+  auto err = cudaGetDevice(&tmp_device);
+  if (err == cudaSuccess) {
+    TORCH_INTERNAL_ASSERT(
+        tmp_device >= 0 &&
+            tmp_device <= std::numeric_limits<DeviceIndex>::max(),
+        "cudaGetDevice returns invalid device ",
+        tmp_device);
```

Comment:
I'm not sure it's worth doing the extra error test here, from a micro optimization perspective. We should probably just trust that cudaGetDevice isn't going to give us garbage when cudaSuccess